app/src/lib/stores/app-store.ts

```diff
@@ -2045,12 +2077,126 @@ export class AppStore extends TypedBaseStore<IAppState> {
     const workingDirectory = WorkingDirectoryStatus.fromFiles(updatedFiles)
 
     this.repositoryStateCache.updateChangesState(repository, () => ({
-      diff,
+      selection: {
+        ...changesState.selection,
+        diff,
+      },
       workingDirectory,
     }))
     this.emitUpdate()
   }
 
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public async _selectStashedFile(
+    repository: Repository,
+    file?: CommittedFileChange | null
+  ): Promise<void> {
+    this.repositoryStateCache.updateChangesState(repository, state => {
+      let selectedStashedFile: CommittedFileChange | null
+
+      if (file === undefined) {
+        if (
+          state.stashEntry &&
+          state.stashEntry.files.kind === StashedChangesLoadStates.Loaded &&
+          state.stashEntry.files.files.length > 0
+        ) {
+          selectedStashedFile = state.stashEntry.files.files[0]
+        } else {
+          selectedStashedFile = null
+        }
+      } else {
+        const { stashEntry } = state
+        if (
+          file !== null &&
+          stashEntry !== null &&
+          stashEntry.files.kind === StashedChangesLoadStates.Loaded
+        ) {
+          const files = stashEntry.files.files
+
+          // Look up the selected file in the stash entry, it's possible that
+          // the stash entry or file list has changed since the consumer called
+          // us. The workingdirectory selection handles this by using IDs rather
+          // than references.
+          selectedStashedFile = files.find(x => x.id === file.id) || null
+        } else {
+          selectedStashedFile = null
+        }
+      }
+
+      return {
+        selection: {
+          kind: ChangesSelectionKind.Stash as ChangesSelectionKind.Stash,
```

Comment:
This, and the other similar casts in here is a concession I had to make to get around a TypeScript compiler error that I haven't had the time to get to the bottom of.

```Type 'ChangesSelectionKind' is not assignable to type 'ChangesSelectionKind.Stash'```

I don't get this error in vs code fwiw.

Reply:
> I don't get this error in vs code fwiw.

Oh, I should have realized that's because I had vs code configured to run the typescript version that ships with vs code and not the one configured for our project. So this will be solved the next time we bump typescript most likely.